mws/apis/feeds.py

```diff
@@ -1,45 +1,57 @@
-from mws.decorators import next_token_action
+import string
 
+from mws import MWS
+from mws.decorators import next_token_action
+from mws.utils.crypto import calc_md5
+from mws.utils.params import enumerate_param
 
 # TODO Add FeedProcessingStatus enumeration
 # TODO Add FeedType enumeration
 
 
-def feed_options_str(feed_options):
-    """Convert a FeedOptions dict of values into an appropriate string value.
+def clean_feed_option_val(val):
+    """Cleans ``val`` for use in the FeedOptions parameter when submitting a feed
+    for "_UPLOAD_VAT_INVOICE_" feed type.
+
+    Amazon states the only safe characters are:
+
+        ``,`` (commas), ``/`` and ``\\`` (slashes), ``-`` (dash), ``_`` (underscore),
+        ``;`` (semi colon), ``:`` (colon), ``#``, 0-9, A-Z, a-z, spaces
+
+    Any character not matching the above set is stripped.
+    """
+    permitted = string.ascii_letters + string.digits + ",\\/-_;:#" + " "
+    # Note the intentional space character added at the end for clarity!
+    return "".join(c for c in val if c in permitted)
 
-    Amazon docs for VAT upload with details:
-    https://m.media-amazon.com/images/G/01/B2B/DeveloperGuide/vat_calculation_service__dev_guide_H383rf73k4hsu1TYRH139kk134yzs.pdf
-    (section 6.4)
+
+def feed_options_str(feed_options):
+    """Convert a ``feed_options`` dict into a formatted metadata string,
+    for use with the SubmitFeed ``FeedOptions`` parameter when submitting
+    VAT invoices.
 
     Example:
-      feed_options = {
-        "shippingid": "283845474",
-        "totalAmount": 3.25,
-        "totalvatamount": 1.23,
-        "invoicenumber": "INT-3431-XJE3",
-        "documenttype": "CreditNote",
-        "transactionid": "amzn:crow:429491192ksjfhe39s",
-      }
-      print(feed_options_str(feed_options))
-      >>> "metadata:shippingid=283845474;metadata:totalAmount=3.25;metadata:totalvatamount=1.23;
-      metadata:invoicenumber=INT-3431-XJE3;metadata:documenttype=CreditNote;
-      metadata:transactionid=amzn:crow:429491192ksjfhe39s"
+
+    .. code-block:: python
+
+        feed_opts = {'orderid': '407-XXXXXX-6760332', 'invoicenumber': 51}
+        opts_str = feed_options_str(feed_opts)
+        # 'metadata:orderid=407-XXXXXX-6760332;metadata:invoicenumber=51'
+
+    See Amazon documentation, `Invoice Uploader Developer Guide (PDF)
+    <https://m.media-amazon.com/images/G/03/B2B/invoice-uploader-developer-documentation.pdf>`_
+    (section 6.4), for details.
     """
     if not feed_options:
         return None
     if not isinstance(feed_options, dict):
         raise ValueError("`feed_options` should be a dict or None")
     output = []
     for key, val in feed_options.items():
-        outval = val
-        if outval is True or outval is False:
-            # Convert literal `True` or `False` to strings `"true"` and `"false"`
-            outval = str(outval).lower()
-        output.append("metadata:{}={}".format(key, outval))
+        clean_val = clean_feed_option_val(val)
```

Comment:
In the old version you were explicitly checking if the `outval` was `True` or `False` then turning into a lowercase string.

I'm making the assumption this is because of the change you made for force using Python native Booleans (mentioned in your changes doc and therefor we don't need to convert to a string?

Mentioning so I don't forget, as I've not got to that part of the code yet.

Reply:
Good catch, that was a shortcoming. I had also not designed it properly to handle numbers, despite the docstring saying it could.

It should now be able to handle bools, numbers, datetimes, and None, similar to other cleaning methods (`None` as a sentinel value to exclude a key-value pair; though this will need to be documented in more detail later).